File: macro_2/lista_2/main_script.m
clc
clear
close all

%tic

% Main Script
% 1. Parametrizacao da Economia

sigma = 1.5;
beta = 0.9932;
q = 1;
a_min = -4;
a_max = 6;
N = 300;
y_L = 0.1;
y_H = 1.0;

% Probabilidades de Transicao
pi_HH = 0.925;
pi_HL = 0.5;
pi_LH = 1 - pi_HH;
pi_LL = 1 - pi_HL;

% Grids
a_grid = linspace(a_min, a_max, N);
y_grid = [y_L, y_H];
pi = [pi_LL, pi_LH; pi_HL, pi_HH]';

M = length( y_grid );

V = zeros( N,M );

% 2. Primeira Etapa Chutes Iniciais

[ V, a_pol ] = vfi(beta,sigma,y_grid,a_grid,q,pi,M,N,V);

% Calculo Distribuicao e excesso de demanda

[ stat_dist, d_excess ] = demand_n_distr( a_pol, a_grid, pi, N, M);

% 3. Etapa Encontrado valor para inciar Bissecao

% Dado excesso de demanda C vamos ver o que fazer:

[ q1, d_excess_1 ] = find_q(d_excess,q,y_grid,beta,sigma,a_grid,pi,V,M,N,0.1 );

% 4. Aplicar Bicessao e encontrar nosso q otimo

% Parametrizando Iteracoes
iter_b     = 0;
max_iter_b = 100;
tol_b      = 1e-5;

% Organizando Precos
if q1 > q
    qb         = q1;
    qa         = q;
    d_excess_a = d_excess;
    d_excess_b = d_excess_1;
else
    qa         = q1;
    qb         = q;
    d_excess_a = d_excess_1;
    d_excess_b = d_excess;
end

hist_q = [ q q1 ];
hist_d = [ d_excess d_excess_1 ];

% Bisection method loop
while (qb - qa) / 2 > tol_b && iter_b < max_iter_b
    qc = (qa + qb) / 2;

    % Iterando funcao valor com novo preco
    [ V, a_pol ] = vfi(beta,sigma,y_grid,a_grid,qc,pi,M,N,V);

    % Calculando Excesso de Demanda
    [ stat_dist_c, d_excess_c ] = demand_n_distr( a_pol, a_grid, pi, N, M );
   
    % Atualizando
    if d_excess_c == 0
        break;
    elseif d_excess_a * d_excess_c < 0
        qb         = qc;
        d_excess_b = d_excess_c;
    else
        qa         = qc;
        d_excess_a = d_excess_c;
    end
    % Atualiza Iteracao
    iter_b = iter_b + 1;

    % Guarda Historico
    hist_d(end+1) = d_excess_c;
    hist_q(end+1) = qc;
    
    % Display iteracao
    disp( iter_b   );
    disp( d_excess_c );
end

toc

% Plots - 

% Funcao Valor Convergia
plot( a_grid, V_conv )

% Funcao Politica
figure(1)
plot( a_grid, a_pol(:,1) );
hold on;
plot( a_grid, a_pol(:,1) );
plot( a_grid, a_grid );
hold off ;
legend( 'High', 'Low', '90 Graus');

% Distribuicao
plot( sum( stat_dist_c,2) )


% Distribuicao por heterogeneidade
figure(2)
plot( a_grid, stat_dist(:,1) )
hold on
plot( a_grid, stat_dist(:,2) )
xlabel('Assets (a)');
ylabel('Density');
hold off
legend('High','Low')


% Scatter Plot - Excesso de Demanda e Precos
scatter( hist_q, hist_d)
xlabel( 'q')

File: macro_2/lista_2/Python/main_script.py
# Importanto Bibliotecas
import numpy as np
import numba as nb
import matplotlib.pyplot as plt
import time
import sys
import os

# ATENÇÃO: Selecionar a raiz
path = r'C:\\Users\\...'

sys.path.append( path )
# Importanto Funcoes
import functions as fc

t0 = time.time()
# Mains Script
sigma = 1.5
beta  = .9932
q     = 1
a_min = -4
a_max = 6
n     = 300

# Vator de Estados
X     = np.array([ 0.1, 1.0 ])

# Matriz de transicao
pi    = np.array([ [ .5  , .5   ],
                   [ .075, .925 ] ])

# Funcao Anonima da Utilidade
U = lambda c: (c**(1-sigma))/(1-sigma)

def U(c,sigma=1.5):
    return ((c**(1-sigma))/(1-sigma))

# Construção dos Grids
a_grid = np.linspace( a_min, a_max, n )
y_grid = X

# 2. Dado parametros iniciais Calcula
dist, Vc, pol_p, pol_a = fc.vfi( beta, y_grid, a_grid, q, pi, U )

s_d, de = fc.demand_n_distr( pol_a, a_grid, pi ) 

# 3. Buscando Outro Preco para poder realizar Bissecao

q1, de1 = fc.find_q(de, q, y_grid, beta, a_grid, pi, U, .1 )

# 4. Iniciar Bissecao para encontrar preco que zera Demanda Agregada

# Parametrizando Iteracoes
iter_b     = 0
max_iter_b = 100
tol_b      = 1e-5

if q1 > q:
    qb = q1
    qa = q
    da = de
    db = de1
else:
    qa = q1
    qb = q
    da = de1
    db = de

# Historicos
hist_q = [ q, q1 ]
hist_d = [ de, de1 ]

while (qb-qa)/2 > tol_b and iter_b < max_iter_b :
    # Ponto Intermediario
    qc = (qa+qb)/2
    # Recalculando VFI
    dist_1, Vc_1, pol_p_1, pol_a_1 = fc.vfi( beta, y_grid, a_grid, qc, pi, U )
    # Calcula Distr Estac e Excesso de demanda
    s_e, dc = fc.demand_n_distr( pol_a_1, a_grid, pi ) 
    # Criterio
    if dc == 0:
        break
    elif da*dc < 0:
        qb = qc
        db = dc
    else:
        qa = qc
        da = dc
    # Atualiza iteracao e Hist
    iter_b = iter_b + 1
    hist_d.append( dc )
    hist_q.append( qc )
    # Print Resultado
    print( qc, dc, iter_b)

t1 = time.time()

print( t1-t0)
# 5. Plots

np.save( path+r'results'+str(round(abs(a_min)))+r'.npy', np.array([ qc, Vc_1, pol_a_1, s_e, dc ]))


# Funcao Valor Convergida
plt.plot( a_grid, Vc[:,0], label = 'Low' )
plt.plot( a_grid, Vc[:,1], label = 'High' )
plt.xlabel(r'$a$')
plt.ylabel(r'$V$')
plt.legend()
#plt.savefig(path+r'figs\\plot_1.pdf')
plt.show()
plt.close()

# Funcao Politica
plt.plot( a_grid, pol_a[:,0], label = 'Low' )
plt.plot( a_grid, pol_a[:,1], label = 'High' )
plt.plot( a_grid, a_grid, linestyle = '--', color = 'grey', label = '90 graus' )
plt.legend()
plt.ylabel(r"$a'$")
plt.xlabel(r"$a$")
#plt.savefig(path+r'figs\\plot_2.pdf')
plt.show()
plt.close()

# Distribuicao Convergida
plt.plot( a_grid, s_d[0,:], label = 'Low' )
plt.plot( a_grid, s_d[1,:], label = 'High' )
plt.ylabel('Percent')
plt.xlabel( r'$a$')
plt.legend()
#plt.savefig(path+r'figs\\plot_3.pdf')
plt.show()
plt.close()

# Scatterplot de Precos e Excesso de Demanda
plt.scatter( hist_q, hist_d )
plt.xlabel(r'$q$')
plt.ylabel(r'Excesso de Demanda')
#plt.savefig(path+r'figs\\plot_4.pdf')
plt.show()
plt.close()

t1 = time.time()

print( t1-t0,' Seconds')


# (b) Achar a matriz M (a,y) 

# Número de Pontos

npo, ns = np.shape( pol_a )
# Vetor Nulo
Va = []
# Loop preenchimento de matrizes
for k in range(0, ns ):
    Va.append( pol_a[:,k][:,np.newaxis] == a_grid)
# Convertendo para Array
Va = np.array( Va, dtype=int  )
# Criando Matriz M
M = np.zeros( (npo*ns,npo*ns) )
# Preenchendo matriz M
for i in range(0,ns):
    for j in range(0,ns):
        #print( i*npo,':',(i+1)*(npo-1), '   ', j*npo,':',(j+1)*(npo-1) )
        M[ i*(npo):(i+1)*(npo) , j*(npo):(j+1)*(npo) ] = Va[i]*pi[i,j]

# Olhando o Plot da Matrix M

plt.imshow(M)
plt.colorbar()
plt.show()

auto_val, auto_vecs = np.linalg.eig( np.matrix(M).T)

idx = np.argmin(np.abs(auto_val - 1))

stationary = np.real(auto_vecs[:, idx])

stationary = stationary / stationary.sum()

stat_reshape = np.reshape( stationary, (ns,npo) ).T

stationary_2 = np.matrix(M)**1000

stat_reshape_2 = np.reshape( np.array(stationary_2 )[1,:], (ns,npo)).T

plt.plot( stat_reshape, "--" ,label = "Metodo 1" )
plt.plot( stat_reshape_2, label = "Metodo 2" )
plt.legend()
plt.show()
